tests/integration/integration_test.go

```diff
@@ -234,6 +235,73 @@ func TestRemoveWithResourcesBlocked(t *testing.T) {
 	e.RunCommand("pulumi", "stack", "rm", "--yes")
 }
 
+// TestPreviewJSON tests serializing a preview as JSON using the `pulumi preview --json` option.
+func TestPreviewJSON(t *testing.T) {
+	if os.Getenv("PULUMI_ACCESS_TOKEN") == "" {
+		t.Skipf("Skipping: PULUMI_ACCESS_TOKEN is not set")
+	}
+
+	e := ptesting.NewEnvironment(t)
+	defer func() {
+		if !t.Failed() {
+			e.DeleteEnvironment()
+		}
+	}()
+
+	stackName, err := resource.NewUniqueHex("rm-test-", 8, -1)
+	contract.AssertNoErrorf(err, "resource.NewUniqueHex sould not fail with no maximum length is set")
+
+	e.ImportDirectory("single_resource")
+	e.RunCommand("pulumi", "stack", "init", stackName)
+	e.RunCommand("yarn", "link", "@pulumi/pulumi")
+
+	// Run the preview to get the JSON. Trim out some pieces we don't want to be sensitive to (like
+	// the precise details of dynamic provider serialization), and then ensure the text matches what we expect.
+	stdout, _ := e.RunCommand("pulumi", "preview", "--json")
+	// nolint: lll
+	expect := fmt.Sprintf(`{
+    "steps": [
+        {
+            "op": "create",
+            "urn": "urn:pulumi:%[1]s::protect_resources::pulumi:pulumi:Stack::protect_resources-%[1]s",
+            "newState": {
+                "urn": "urn:pulumi:%[1]s::protect_resources::pulumi:pulumi:Stack::protect_resources-%[1]s",
+                "custom": false,
+                "type": "pulumi:pulumi:Stack"
+            }
+        },
+        {
+            "op": "create",
+            "urn": "urn:pulumi:%[1]s::protect_resources::pulumi-nodejs:dynamic:Resource::res",
+            "provider": "urn:pulumi:%[1]s::protect_resources::pulumi:providers:pulumi-nodejs::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+            "newState": {
+                "urn": "urn:pulumi:%[1]s::protect_resources::pulumi-nodejs:dynamic:Resource::res",
+                "custom": true,
+                "type": "pulumi-nodejs:dynamic:Resource",
+                "inputs": {
+                    "__provider": "...",
+                    "state": 1
+                },
+                "parent": "urn:pulumi:%[1]s::protect_resources::pulumi:pulumi:Stack::protect_resources-%[1]s",
+                "provider": "urn:pulumi:%[1]s::protect_resources::pulumi:providers:pulumi-nodejs::default::04da6b54-80e4-46f7-96ec-b56ff0331ba9",
+                "propertyDependencies": {
+                    "__provider": null,
+                    "state": null
+                }
+            }
+        }
+    ],
+    "changeSummary": {
+        "create": 2
+    }
```

Comment:
primary concern here is: are we signing up for this being a format we will always support?  if we cahnge our eventing structure, do we have a compat issue here?

Reply:
Yes, we are, as I assume to be the case for any data structure we serialize to JSON. This is primarily leveraging existing data structures that we serialize in the `apitype` namespace.